### PR TITLE
normalization_run

### PR DESCRIPTION
**Pull request by cansavvy**
This PR is focused on a script that runs single cell rna-seq normalizations from command line. 

The `asap_data_prep.Rmd`s and `run_pipeline.sh` also have some minor changes to fit with this, but the main adds are the `5-run_normalizations.R` script which was adapted from this [ASAP script](https://github.com/DeplanckeLab/ASAP/blob/master/R_Python/normalization.R)

#### Options:
```
'-d' : Path to gene matrix in tab delimited format, gene x sample with gene 
        info as the first column 
'-a' : Normalization method to use. To run all methods use "all", to run more
        than one method, separate the names by a space. eg -a tmm log
        Options:  
            'scale'- scale and center each sample's data using base::scale 
           'log'- just log transform alone . 
           'voom'- quantile normalization from the DEseq2::voom function
           'deseq2'- Use DEseq2 size factors
           'vsd'- Use DEseq2 varianceStabilizingTransformation function
           'rld'- Use DEseq2 rlogTransformation function
           'tmm'- trimmed mean M values by edgeR
           'all'- run all the above methods
 '-o' : Directory where you would like the output to go. Default is current
        directory
 '-l' : Optional label for output files
```
 #### Command line example:
 ```bash
Rscript scripts/5-run_normalization.R \
 -d data/darmanis_counts.tsv \
 -a log \
 -o normalized_data \
 -l darmanis
```

### CODE REVIEW

**Comment by cansavvy** [comment]
Unless you have any particular objections, I'm going to remove`DESeq2::rlogTransformation` as an option. It takes too long. Not practical.

**Comment by jaclyn-taroni** [comment]
No objections, it should be similar enough to VST

**Comment by jaclyn-taroni** [comment]
Looks like that got removed already (not sure if you have any other changes coming), can you also make sure:
- Comments are up to date
- Style is consistent with the guide we use and the rest of the repo (e.g., spacing, indentation)
- Any URLs to code references use a relevant commit hash and don't link to whatever the most recent version is

please and then let me know and I will take a 👀

**Comment by jaclyn-taroni** [comment]
What's going on with the spacing in https://github.com/cansav09/scRNA-seq_workflow/pull/14/files#diff-916e9ccc58fe4036c569a8a743ea8c12?

File: run_pipeline.sh [hunk 1]
```diff
@@ -79,6 +84,17 @@ Rscript scripts/2-get_fastqc_reports.R \
 Rscript scripts/3-make_gene_matrix.R \
   -d ${dir}/salmon_quants \
   -o ${dir} \
-  -g ${GSE} \
   -m 0.5 \
   -l ${dir}
+
+#-------------------------------Get metadata tsv-------------------------------#
+Rscript scripts/4-metadata_setup.R \
```
Comment (jaclyn-taroni) [inline run_pipeline.sh]: Where is `scripts/4-metadata_setup.R` coming from?

File: scripts/5-run_normalization.R [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+#--------------------------------Set up options--------------------------------#
+# Set up optparse options
+option_list <- list(
+  make_option(opt_str = c("-d", "--data"), type = "character", default = "none",
+              help = "Path to gene matrix in tab delimited format, gene x sample
+              with gene info as the first column",
+              metavar = "character"),
+  make_option(opt_str = c("-a", "--algorithm"), type = "character",
+              default = "none", help = "Normalization method to use. Options:
+              'scale', log', 'voom', 'deseq2', 'vsd', 'rld', 'tmm' or use 'all'
```
Comment (jaclyn-taroni) [inline scripts/5-run_normalization.R]: You took out `rld`, correct?

File: scripts/5-run_normalization.R [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+# Stop if no input data matrix is specified
+if (opt$data == "none") {
+    stop("Error: no specified input gene matrix file. Use option -d to specify
+         the gene matrix tsv data you would like to normalize")
+}
+
+# Stop if no normalization algorithm is specified
+if (opt$algorithm == "none") {
+  stop("Error: no normalization algorithm was chosen. Use option -a to choose. 
+       Options include:'scale', 'log', 'voom','tmm','deseq2', 'vsd', 'rld' or
```
Comment (jaclyn-taroni) [inline scripts/5-run_normalization.R]: Same comment about `rld`

File: scripts/5-run_normalization.R [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+} else {
+    # Separate if multiple algorithms have been requested
+    opt$algorithm <- unlist(strsplit(opt$algorithm, " "))
+}
+
+# Check that the algorithms chosen are in the supported list
+if (any(is.na(match(opt$algorithm, all.algorithms)))) {
+    stop("That is not a normalization method supported by this script.
+         Check for typos. Acceptable options:'scale', 'log', 'voom','tmm',
+         'deseq2', 'vsd', 'rld' or 'all")
```
Comment (jaclyn-taroni) [inline scripts/5-run_normalization.R]: Same comment about `rld`

File: scripts/5-run_normalization.R (hunk 3 above)
Comment (jaclyn-taroni) [reply scripts/5-run_normalization.R]: Do you need to do this part or can you just let the check at line 94 catch this case as well?

File: scripts/5-run_normalization.R [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+}
+#----------------------------------Load data-----------------------------------#
+# Read in a tsv file of data
+dataset <- readr::read_tsv(opt$data)
+
+# Separate genes from the numeric data
+genes <- dataset[, grep("gene", colnames(dataset), ignore.case = TRUE)]
```
Comment (jaclyn-taroni) [inline scripts/5-run_normalization.R]: Looks like it's a requirement that the column that contains gene identifiers is named `"gene"`, you should document this and perhaps use `stop()` if there's no `"gene"` column

File: scripts/5-run_normalization.R [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+                    DESeq2::varianceStabilizingTransformation(data.dds, 
+                                                              blind = TRUE))
+        title <- "Log2 Expression (varianceStabilizingTransformation - DEseq2)"
+        
+    }
+    # Calculate percent zeroes
+    perc.zero <- round(length(which(data.out == 0)) / length(as.matrix(data.out)), 3)
+    
+    # Build output file name:
+    output.file <- paste0(opt$output, "/", algorithm, "_", opt$label, ".tsv")
```
Comment (jaclyn-taroni) [inline scripts/5-run_normalization.R]: Couple things here:
- You should use `file.path()` for the path separator bit.
- If I were to not specify a label, the output file name would be (for log): `log_.tsv`, is that correct? Seems kind of ugly.

File: arnon_data_prep.Rmd [hunk 7]
```diff
@@ -70,7 +73,12 @@ readr::write_tsv(prepped.data, file.path("arnon_data", "normalized_arnon",
 readr::write_tsv(data.frame(batch.info), file.path("arnon_data",
                                                    "batch.info.txt"))
 ```
-
+### Write metadata info to be used later
+```{r Write batch file}
+# Write metadata file for subsequent testing
+geo.meta %>% dplyr::select("cell.type.ch1", "characteristics_ch1.2") %>%
```
Comment (jaclyn-taroni) [inline arnon_data_prep.Rmd:79]: Difference from what you have earlier is that you're only saving selected metadata, do I have that right?

File: scripts/5-run_normalization.R [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+
+# Create the output for results folder if it does not exist
+if (!dir.exists(opt$output)) {
+    message(paste("Output folder:", opt$output, "does not exist, creating one"))
+    dir.create(opt$output)
+}
+
+#---------------------------Set up algorithms to run---------------------------#
+# List all the supported algorithms
+all.algorithms <- c( 'scale', 'log', 'voom','tmm', 'deseq2', 'vsd', 'rld')
```
Comment (jaclyn-taroni) [inline scripts/5-run_normalization.R]:
```suggestion
all.algorithms <- c('scale', 'log', 'voom', 'tmm', 'deseq2', 'vsd')
```

**Review by jaclyn-taroni — APPROVED** [review]
Looks like you missed some references to `rld` in https://github.com/cansav09/scRNA-seq_workflow/pull/14/commits/dd33912331caf8201578e97806dc92e8efbdcf29 -- in the future, it's good practice to make sure all fixes are included in the same commit.

File: run_pipeline.sh (hunk 1 above)
Comment (cansavvy) [reply run_pipeline.sh]: It's outdated, I thought maybe I should bring it back, but I think this step is better served in an Rmd.

File: scripts/5-run_normalization.R (hunk 2 above)
Comment (cansavvy) [reply scripts/5-run_normalization.R]: Yes, I thought I took out all of these, but I underestimated how many points I reference "rld".

File: arnon_data_prep.Rmd (hunk 7 above)
Comment (cansavvy) [reply arnon_data_prep.Rmd:79]: Yes. Saving only the metadata we want to test/label

File: scripts/5-run_normalization.R (hunk 3 above)
Comment (cansavvy) [reply scripts/5-run_normalization.R]: I changed it so that it does this checking all in one step.

**Comment by cansavvy** [comment]
Okay. I believe I've addressed all the suggested changes

**Comment by jaclyn-taroni** [comment]
Yep, you're approved so feel free to merge 👍

**Pull request merged**